File: 프로그래머스/1/67256. ［카카오 인턴］ 키패드 누르기/［카카오 인턴］ 키패드 누르기.js
function solution(numbers, hand) {
    let key = [['1', '4', '7', '*'], ['2', '5', '8', '0'], ['3', '6', '9', '#']];
    let hands = {
        left: '*',
        right: '#'
    }
    let result = '';
    numbers.map(i => String(i)).forEach((item) => {
        let lefthand = 0
        let righthand = 0
        if(key[0].includes(item)){
            hands.left = item;
            result += 'L';
        } else if(key[2].includes(item)){
            hands.right = item;
            result += 'R';
        } else{
            if(key[0].includes(hands.left)){
                lefthand = Math.abs(key[0].indexOf(hands.left) - key[1].indexOf(item))+1
            } else{
                lefthand = Math.abs(key[1].indexOf(hands.left) - key[1].indexOf(item))
            }
            if(key[2].includes(hands.right)){
                righthand = Math.abs(key[2].indexOf(hands.right) - key[1].indexOf(item))+1
            } else{
                righthand = Math.abs(key[1].indexOf(hands.right) - key[1].indexOf(item))
            }
            
            if(lefthand > righthand){
                result += 'R';
                hands.right = item;
            } else if(lefthand < righthand){
                result += 'L';
                hands.left = item;
            } else{
                if(hand === "right"){
                    result += 'R'
                    hands.right = item
                } else{
                    result += 'L'
                    hands.left = item
                }
            }
        }
    })
    return result
}

File: 프로그래머스/1/67256. ［카카오 인턴］ 키패드 누르기/［카카오 인턴］ 키패드 누르기.py
def solution(numbers, hand):
    answer = ""

    keypad = [
        [1, 2, 3],
        [4, 5, 6],
        [7, 8, 9],
        ["*", 0, "#"]
    ]

    left = "*"
    right = "#"

    for number in numbers:

        if number in [1, 4, 7]:
            answer += "L"
            left = number

        elif number in [3, 6, 9]:
            answer += "R"
            right = number

        else:
            # 왼손 위치 찾기
            for i in range(4):
                for j in range(3):
                    if keypad[i][j] == left:
                        left_x = i
                        left_y = j

            # 오른손 위치 찾기
            for i in range(4):
                for j in range(3):
                    if keypad[i][j] == right:
                        right_x = i
                        right_y = j

            # 목표 숫자 위치 찾기
            for i in range(4):
                for j in range(3):
                    if keypad[i][j] == number:
                        target_x = i
                        target_y = j

            left_distance = abs(left_x - target_x) + abs(left_y - target_y)
            right_distance = abs(right_x - target_x) + abs(right_y - target_y)

            if left_distance < right_distance:
                answer += "L"
                left = number

            elif right_distance < left_distance:
                answer += "R"
                right = number

            else:
                if hand == "left":
                    answer += "L"
                    left = number
                else:
                    answer += "R"
                    right = number

    return answer